src/Microsoft.Graph.Core/Tasks/LargeFileUploadTask.cs

```diff
@@ -0,0 +1,258 @@
+                        this.TotalUploadLength);
+
+                    yield return uploadRequest;
+
+                    currentRangeBegins += nextSliceSize;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Upload the whole session.
+        /// </summary>
+        /// <param name="maxTries">Number of times to retry entire session before giving up.</param>
+        /// <param name="options">Query and header option name value pairs for the request.</param>
+        /// <param name="progress">IProgress object to monitor the progress of the upload.</param>
+        /// <returns>Item information returned by server.</returns>
+        public async Task<UploadResult<T>> UploadAsync(IProgress<long> progress = null, int maxTries = 3, IEnumerable<Option> options = null)
+        {
+            var uploadTries = 0;
+            var trackedExceptions = new List<Exception>();
+
+            while (uploadTries < maxTries)
+            {
+                var sliceRequests = this.GetUploadSliceRequests(options);
+
+                foreach (var request in sliceRequests)
+                {
+                    var uploadResult = await this.UploadSliceAsync(request, trackedExceptions).ConfigureAwait(false);
+                    
+                    progress?.Report(request.RangeBegin);//report the progress of upload
+
+                    if (uploadResult.UploadSucceeded)
+                    {
+                        return uploadResult;
+                    }
+                }
+
+                await this.UpdateSessionStatusAsync().ConfigureAwait(false);
+                uploadTries += 1;
+                if (uploadTries < maxTries)
+                {
+                    // Exponential back off in case of failures.
+                    await Task.Delay(2000 * uploadTries * uploadTries).ConfigureAwait(false);
+                }
+            }
+            
+            throw new TaskCanceledException("Upload failed too many times. See InnerException for list of exceptions that occured.", new AggregateException(trackedExceptions.ToArray()));
+        }
+
+        /// <summary>
+        /// Get info about the upload session and resume from where it left off.
+        /// </summary>
+        /// <param name="maxTries">Number of times to retry entire session before giving up.</param>
+        /// <param name="options">Query and header option name value pairs for the request.</param>
+        /// <param name="progress">IProgress object to monitor the progress of the upload.</param>
+        /// <returns>Item information returned by server.</returns>
+        public async Task<UploadResult<T>> ResumeAsync(IProgress<long> progress = null, int maxTries = 3, IEnumerable<Option> options = null)
+        {
+            var uploadSession = await this.UpdateSessionStatusAsync().ConfigureAwait(false);
+            var uploadExpirationTime = uploadSession.ExpirationDateTime ?? DateTimeOffset.Now;
```

Comment:
How can an uploadSession not have an ExpirationDateTime?

Reply:
As the `ExpirationDateTime` field is nullable from the [Service Library](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/d3aac1579a2ff97c2ce0d6543e816e3dc9ce4913/src/Microsoft.Graph/Models/Generated/UploadSession.cs#L38), we are just making sure that such a scenario is catered for by assuming the session is expired by setting a time that is expired.